File: backend/api/stream.py
"""Server-Sent Events: новые сообщения и результаты оценки приезжают на фронт.
WebSocket не нужен — поток односторонний (решение №18)."""
import asyncio
import json

from fastapi import APIRouter, Depends
from fastapi.responses import StreamingResponse

from backend.events import bus
from backend.models import User
from backend.security import current_user

router = APIRouter(prefix="/api", tags=["stream"])

POLL_SECONDS = 1.0
PING_SECONDS = 15.0


async def _event_stream(user_id: int):
    cursor = bus.cursor  # старые события новому подписчику не отдаём
    since_ping = 0.0
    # Первый байт сразу — иначе браузер держит соединение «висящим».
    yield ": ok\n\n"
    while True:
        events, cursor = bus.since(user_id, cursor)
        for event in events:
            payload = json.dumps(event.data, ensure_ascii=False)
            yield f"event: {event.name}\ndata: {payload}\n\n"
        if events:
            since_ping = 0.0
        await asyncio.sleep(POLL_SECONDS)
        since_ping += POLL_SECONDS
        if since_ping >= PING_SECONDS:
            since_ping = 0.0
            yield ": ping\n\n"


@router.get("/stream")
async def stream(user: User = Depends(current_user)) -> StreamingResponse:
    return StreamingResponse(
        _event_stream(user.id),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )

File: backend/internal/routers/stream.go
package routers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// Поток опрашивает шину раз в секунду и раз в 15 секунд шлёт комментарий-пинг,
// чтобы прокси не закрыл соединение как простаивающее.
const (
	pollInterval = time.Second
	pingInterval = 15 * time.Second
)

// handleStream — Server-Sent Events: новые сообщения и результаты оценки
// приезжают на фронт. WebSocket не нужен, поток односторонний (решение №18).
func (s *Server) handleStream(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}
	flusher, ok := w.(http.Flusher)
	if !ok {
		fail(w, http.StatusInternalServerError, "Поток недоступен")
		return
	}

	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")
	w.Header().Set("X-Accel-Buffering", "no")
	w.WriteHeader(http.StatusOK)

	// Первый байт сразу — иначе браузер держит соединение «висящим».
	fmt.Fprint(w, ": ok\n\n")
	flusher.Flush()

	// Старые события новому подписчику не отдаём.
	cursor := s.bus.Cursor()
	sincePing := time.Duration(0)
	ticker := time.NewTicker(pollInterval)
	defer ticker.Stop()

	for {
		select {
		case <-r.Context().Done():
			return
		case <-ticker.C:
			var batch []string
			events, next := s.bus.Since(user.ID, cursor)
			cursor = next
			for _, event := range events {
				payload, err := json.Marshal(event.Data)
				if err != nil {
					continue
				}
				batch = append(batch, fmt.Sprintf("event: %s\ndata: %s\n\n", event.Name, payload))
			}
			if len(batch) > 0 {
				sincePing = 0
				for _, chunk := range batch {
					fmt.Fprint(w, chunk)
				}
				flusher.Flush()
				continue
			}
			sincePing += pollInterval
			if sincePing >= pingInterval {
				sincePing = 0
				fmt.Fprint(w, ": ping\n\n")
				flusher.Flush()
			}
		}
	}
}
